File: src/sample/java/chatroom/Chatroom.java
package chatroom;

import com.google.gson.Gson;
import webbit.WebSocketConnection;
import webbit.WebSocketHandler;

import java.util.HashMap;
import java.util.Map;

public class Chatroom implements WebSocketHandler {

    private final Gson json = new Gson();

    static class Incoming {
        enum Action { LOGIN, SAY }
        Action action;
        String loginUsername;
        String message;
    }

    static class Outgoing {
        enum Action { JOIN, LEAVE, SAY }
        Action action;
        String username;
        String message;
    }

    private Map<WebSocketConnection, String> usernames = new HashMap<WebSocketConnection, String>();

    @Override
    public void onOpen(WebSocketConnection connection) throws Exception {
        // Don't do anything until we receive a LOGIN message.
    }

    @Override
    public void onMessage(WebSocketConnection connection, String msg) throws Exception {
        Incoming incoming = json.fromJson(msg, Incoming.class);
        switch (incoming.action) {
            case LOGIN:
                login(connection, incoming.loginUsername);
                break;
            case SAY:
                say(connection, incoming.message);
                break;
        }
    }

    private void login(WebSocketConnection connection, String username) {
        usernames.put(connection, username);
        Outgoing outgoing = new Outgoing();
        outgoing.action = Outgoing.Action.JOIN;
        outgoing.username = username;
        broadcast(outgoing);
    }

    private void say(WebSocketConnection connection, String message) {
        String username = usernames.get(connection);
        if (username != null) {
            Outgoing outgoing = new Outgoing();
            outgoing.action = Outgoing.Action.SAY;
            outgoing.username = username;
            outgoing.message = message;
            broadcast(outgoing);
        }
    }

    private void broadcast(Outgoing outgoing) {
        String json = this.json.toJson(outgoing);
        for (WebSocketConnection connection : usernames.keySet()) {
            connection.send(json);
        }
    }

    @Override
    public void onClose(WebSocketConnection connection) throws Exception {
        String username = usernames.get(connection);
        if (username != null) {
            Outgoing outgoing = new Outgoing();
            outgoing.action = Outgoing.Action.LEAVE;
            outgoing.username = username;
            broadcast(outgoing);
            usernames.remove(connection);
        }
    }

    @Override
    public void onError(WebSocketConnection connection, Exception error) throws Exception {
        error.printStackTrace();
    }

}

File: src/test/java/samples/flashchatroom/content/chatroom.js
// web-socket-js 
WEB_SOCKET_SWF_LOCATION = "WebSocketMain.swf";
WEB_SOCKET_DEBUG = true;


// Socket reference.
var ws;

// Log text to main window.
function logText(msg) {
    var textArea = document.getElementById('chatlog');
    textArea.value = textArea.value + msg + '\n';
    textArea.scrollTop = textArea.scrollHeight; // scroll into view
}

// Perform login: Ask user for name, and send message to socket.
function login() {
    var defaultUsername = (window.localStorage && window.localStorage.username) || 'yourname';
    var username = prompt('Choose a username', defaultUsername);
    if (username) {
        if (window.localStorage) { // store in browser localStorage, so we remember next next
            window.localStorage.username = username;
        }
        send({action:'LOGIN', loginUsername:username});
        document.getElementById('entry').focus();
    } else {
        ws.close();
    }
}

function onMessage(incoming) {
    switch (incoming.action) {
        case 'JOIN':
            logText("* User '" + incoming.username + "' joined.");
            break;
        case 'LEAVE':
            logText("* User '" + incoming.username + "' left.");
            break;
        case 'SAY':
            logText("[" + incoming.username + "] " + incoming.message);
            break;
    }
}

// Connect to socket and setup events.
function connect() {
    // clear out any cached content
    document.getElementById('chatlog').value = '';

    // connect to socket
    logText('* Connecting...');
    ws = new WebSocket('ws://' + document.location.host + '/chatsocket');
    ws.onopen = function(e) {
        logText('* Connected!');
        login();
    };
    ws.onclose = function(e) {
        logText('* Disconnected');
    };
    ws.onerror = function(e) {
        logText('* Unexpected error');
    };
    ws.onmessage = function(e) {
        onMessage(JSON.parse(e.data));
    };

    // wire up text input event
    var entry = document.getElementById('entry');
    entry.onkeypress = function(e) {
        if (e.keyCode == 13) { // enter key pressed
            var text = entry.value;
            if (text) {
                send({action:'SAY', message:text});
            }
            entry.value = '';
        }
    };
}

// Send message to server over socket.
function send(outgoing) {
    ws.send(JSON.stringify(outgoing));
}

// Connect on load.
window.onload = connect;
